File: CodeForces/F_Money_Trees.cpp
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
int n;
ll h;
vector<ll> fruits, height;

void solve()
{
    cin >> n >> h;

    fruits.resize(n);
    height.resize(n);

    for (auto &x : fruits)
        cin >> x;
    for (auto &x : height)
        cin >> x;

    ll sum = 0;
    int l = 0;
    int ans = 0;

    for (int r = 0; r < n; r++)
    {
        // break divisibility chain
        if (r > 0 && height[r - 1] % height[r] != 0)
        {
            sum = 0;
            l = r;
        }

        sum += fruits[r];

        while (sum > h)
        {
            sum -= fruits[l];
            l++;
        }

        ans = max(ans, r - l + 1);
    }

    cout << ans << "\n";
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int t;
    cin >> t;
    while (t--)
    {
        solve();
    }
    return 0;
}

File: CodeForces/F_Money_Trees.py
t = int(input())
for _ in range(t):
    n,x = map(int , input().split())
    a = list(map(int , input().split()))
    h = list(map(int , input().split()))
    
    ans = 0
    l = 0
    f = 0
    
    for r in range(n):
        if( r > 0 and (h[r-1] % h[r] != 0)):
            f = 0
            l = r
        
        f += a[r]
        while(f > x):
            f -= a[l]
            l+=1        
    
        ans = max(ans , r-l+1)
        
    print(ans)